### PR TITLE
fix generate config for ~/ relative paths

### PR DESCRIPTION
**Pull request by rtpascual**
## Problem

Running something like `amplify generate config --out-dir=~/test` creates a `~` folder at the current working directory which may not be `$HOME/test`.

**Issue number, if available:**

## Changes

If `outDir` starts with `~/` then expand the tilde to be the home directory.

**Corresponding docs PR, if applicable:**

## Validation

Unit test and manually:

Screenshot 2024-04-10 at 4 38 42 PM

## Checklist

- [ ] If this PR includes a functional change to the runtime behavior of the code, I have added or updated automated test coverage for this change.
- [ ] If this PR requires a change to the [Project Architecture README](../PROJECT_ARCHITECTURE.md), I have included that update in this PR.
- [ ] If this PR requires a docs update, I have linked to that docs PR above.
- [ ] If this PR modifies E2E tests, makes changes to resource provisioning, or makes SDK calls, I have run the PR checks with the `run-e2e` label set.

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: packages/client-config/src/paths/get_client_config_path.ts [hunk 1]
```diff
@@ -23,16 +24,18 @@ export const getClientConfigPath = async (
   let targetPath = defaultArgs.out;
 
   if (outDir) {
-    targetPath = path.isAbsolute(outDir)
-      ? outDir
-      : path.resolve(process.cwd(), outDir);
+    // if outDir starts with "~/" or "~$" or "~\" replace tilde with home directory
```
Comment (rtpascual) [inline packages/client-config/src/paths/get_client_config_path.ts:27]:
```suggestion
    // if outDir starts with "~/" or "~\" or outDir is "~", replace tilde with home directory
```

**Comment by rtpascual** [comment]
Any opinions on the "File written" message if we should change or keep this? Currently it's relative to `process.cwd()` so in my example for the PR description we end up with:
```sh
> /Users/rotp/workplace/amplify-backend/packages/cli/lib/amplify.js generate config --stack amplify-testfunction-rotp-sandbox-8844aa0275 --out-dir=~/workplace/test-gen2/test
File written: ../../../test-gen2/test/amplifyconfiguration.json
```

**Comment by rtpascual** [comment]
> 1. There are several commands that accept a "directory/file" input, this expansion will be needed everywhere.
> 2. This won't fix expanding the environment variables such as `$HOME`. Is there a way we can expand this to cover all such use cases? (In shell `echo` can do that, which can be run through node execa as well)

1. So maybe it's worth moving the expansion logic up to the `cli` package to perform the expansion in the yargs handler methods for each command?
2. Makes sense, we can add that logic as well

**Comment by edwardfoyle** [comment]
I think this may be a shell issue rather than something we can solve in the cli itself. I just ran
```sh
npx amplify generate config --stack stack-name --out-dir ~/test-output-location
File written: ../../../test-output-location/amplifyconfiguration.json
```
My shell (zsh) correctly expanded the arg before passing to the command

**Review by edwardfoyle — CHANGES_REQUESTED** [review]
Marking as needs changes (most likely we shouldn't do this at all). see https://github.com/aws-amplify/amplify-backend/pull/1282#issuecomment-2050720146

**Comment by rtpascual** [comment]
> I think this may be a shell issue rather than something we can solve in the CLI itself. I just ran
> 
> ```shell
> npx amplify generate config --stack stack-name --out-dir ~/test-output-location
> File written: ../../../test-output-location/amplifyconfiguration.json
> ```
> 
> My shell (zsh) correctly expanded the arg before passing to the command.
> 
> EDIT: just tested with `--out-dir $HOME/test-location` and that worked as well.

Yeah looks like this was fixed in the latest beta version. With an older version I am getting different results that is the problem this PR is addressing:
```shell
npx -p @aws-amplify/backend-cli@0.12.0-beta.16 amplify generate config --out-dir=~/workplace/test-gen2/test
File written: ~/workplace/test-gen2/test/amplifyconfiguration.json
```

**Comment by rtpascual** [comment]
Closing as this is no longer a bug with the latest [release](https://github.com/aws-amplify/amplify-backend/pull/1240).

**Pull request closed without merging**